### PR TITLE
Handle arrays as an optional argument

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
Towards: https://github.com/lfortran/lfortran/issues/2323

### CODE REVIEW

File: src/libasr/pass/transform_optional_argument_functions.cpp [hunk 1]
```diff
@@ -286,8 +286,21 @@ bool fill_new_args(Vec<ASR::call_arg_t>& new_args, Allocator& al,
...
                 std::string m_arg_i_name = scope->get_unique_name("__libasr_created_variable_");
+                ASR::ttype_t *arg_type = func_arg_j->m_type;
+                if (ASRUtils::is_array(arg_type)) {
+                    ASR::dimension_t* m_dims = nullptr;
+                    size_t n_dims = ASRUtils::extract_dimensions_from_ttype(arg_type, m_dims);
+                    for (size_t i = 0; i < n_dims; i++) {
+                        // Creating an array with size `0`
+                        ASR::ttype_t *int32 = ASRUtils::TYPE(ASR::make_Integer_t(al, x.m_args[i].loc, 4));
+                        m_dims[i].m_length = ASRUtils::EXPR(ASR::make_IntegerConstant_t(al, x.m_args[i].loc, 0, int32));
+                        m_dims[i].m_start = ASRUtils::EXPR(ASR::make_IntegerConstant_t(al, x.m_args[i].loc, 1, int32));
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/transform_optional_argument_functions.cpp:297]: This doesn't seem to be a correct fix. This creates an array with size `0`, if we access this array in the subroutine or function it doesn't seg fault.

File: src/libasr/pass/transform_optional_argument_functions.cpp [hunk 2]
```diff
@@ -286,8 +286,21 @@ bool fill_new_args(Vec<ASR::call_arg_t>& new_args, Allocator& al,
             ASR::Variable_t* func_arg_j = ASRUtils::EXPR2VAR(func->m_args[j]);
             if( x.m_args[i].m_value == nullptr ) {
                 std::string m_arg_i_name = scope->get_unique_name("__libasr_created_variable_");
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/transform_optional_argument_functions.cpp:288]: Here, for the optional argument, we create a dummy variable and pass it as an argument.
For the type, we use the function arg type(`a(:)`), so the seg fault.
What would be the right way to fix this?

Comment (certik) [reply src/libasr/pass/transform_optional_argument_functions.cpp:288]: Why does it segfault if we pass in `a(:)`?

Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/transform_optional_argument_functions.cpp:288]: We are trying to declare the array without the LBound and UBound: `__libasr_created_variable_(:)` and pass this as an argument to the subroutine

**Comment by Thirumalai-Shaktivel** [comment]
The changes seem to fix the issue and it works for fastGPT.
However, I have a concern about an empty array being allocated.

**Comment by certik** [comment]
@czgdp1807 do you know how to finish this one?

It also seems to segfault for fpm at the CI.

**Comment by czgdp1807** [comment]
I am working on #2323. Once I create a PR I will close this one.

**Comment by Thirumalai-Shaktivel** [comment]
Sure, no issues!

**Comment by czgdp1807** [comment]
Closing in favour of https://github.com/lfortran/lfortran/pull/2358

**Pull request closed without merging**